link/uprobe.go

```diff
@@ -0,0 +1,275 @@
+package link
+
+import (
+	"debug/elf"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal"
+	"github.com/cilium/ebpf/internal/unix"
+)
+
+func init() {
+	symbolsCache = internal.NewSymbolsCache()
+}
+
+var (
+	uprobeEventsPath = filepath.Join(tracefsPath, "uprobe_events")
+
+	// reUprobeSymbol is used to strip invalid characters from the uprobe symbol
+	// as they are not allowed to be used as the EVENT token in tracefs.
+	reUprobeSymbol = regexp.MustCompile("[^a-zA-Z0-9]+")
+
+	// TODO(matt): discuss if a cache is needed and if the current implementation
+	// is correct.
+	symbolsCache *internal.SymbolsCache
+)
+
+// Uprobe attaches the given eBPF program to a perf event that fires when the
+// given symbol starts executing in the given binary.
+// For example, /bin/bash::readline():
+//
+//	Uprobe("/bin/bash", "readline", prog)
+//
+// The resulting Link must be Closed during program shutdown to avoid leaking
+// system resources.
+//
+// Since the implementation of Uprobe is similar to the implementation of Kprobe,
+// some functions in this file are not documented thorougly.
+// For a more complete documentation, refer to the Kprobe implementation.
+func Uprobe(path, symbol string, prog *ebpf.Program) (Link, error) {
+	u, err := uprobe(path, symbol, prog, false)
+	if err != nil {
+		return nil, err
+	}
+
+	err = u.attach(prog)
+	if err != nil {
+		u.Close()
+		return nil, err
+	}
+
+	return u, nil
+}
+
+// Uretprobe attaches the given eBPF program to a perf event that fires right
+// before the given symbol exits.
+//
+// The resulting Link must be Closed during program shutdown to avoid leaking
+// system resources.
+func Uretprobe(path, symbol string, prog *ebpf.Program) (Link, error) {
+	u, err := uprobe(path, symbol, prog, true)
+	if err != nil {
+		return nil, err
+	}
+
+	err = u.attach(prog)
+	if err != nil {
+		u.Close()
+		return nil, err
+	}
+
+	return u, nil
+}
+
+// uprobe opens a perf event for the given binary/symbol and attaches prog to it.
+// If ret is true, create a uretprobe.
+func uprobe(path, symbol string, prog *ebpf.Program, ret bool) (*perfEvent, error) {
```

Comment:
This function plus `pmuUprobe` and `tracefsUprobe` seem to duplicate a lot of code around perf event fd creation etc. from their kprobe counterparts. Is it possible to do some refactoring to share more?

Reply:
Agreed, tracefs k/uprobe stuff should share more code.